op-dripper/dripper/driver.go

```diff
@@ -0,0 +1,251 @@
+package dripper
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/ethereum-optimism/optimism/op-dripper/bindings"
+	"github.com/ethereum-optimism/optimism/op-dripper/metrics"
+	"github.com/ethereum-optimism/optimism/op-service/txmgr"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+type Client interface {
+	CodeAt(ctx context.Context, contract common.Address, blockNumber *big.Int) ([]byte, error)
+	CallContract(ctx context.Context, call ethereum.CallMsg, blockNumber *big.Int) ([]byte, error)
+}
+
+type DrippieContract interface {
+	GetDripCount(*bind.CallOpts) (*big.Int, error)
+	Created(*bind.CallOpts, *big.Int) (string, error)
+	Executable(*bind.CallOpts, string) (bool, error)
+}
+
+type DriverSetup struct {
+	Log    log.Logger
+	Metr   metrics.Metricer
+	Cfg    DripExecutorConfig
+	Txmgr  txmgr.TxManager
+	Client Client
+}
+
+type DripExecutor struct {
+	DriverSetup
+
+	wg   sync.WaitGroup
+	done chan struct{}
+
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	mutex   sync.Mutex
+	running bool
+
+	drippieContract DrippieContract
+	drippieABI      *abi.ABI
+}
+
+func NewDripExecutor(setup DriverSetup) (_ *DripExecutor, err error) {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	// Ensure context does't leak.
+	defer func() {
```

Comment:
I don't think you need this unless you're calling DripExecutor from somewhere that will trap the panic. In the event of a panic you don't need to `cancel()` because the program will just crash and exit.